### PR TITLE
azure function integration

### PR DESCRIPTION
**Pull request by gord02**
## What does this PR do?
Creates an integration for the Azure.Functions package to enhance auto-instrumentation.

## Checklist

- [x] The PR description includes an overview of the change
- [x] The PR description articulates the motivation for the change
- [ ] The change includes tests OR the PR description describes a testing strategy
- [x] The PR description notes risks associated with the change, if any
- [x] Newly-added code is easy to change
- [ ] The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- [ ] The change includes or references documentation updates if necessary
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [ ] Title is accurate
- [ ] All changes are related to the pull request's stated goal
- [ ] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [ ] Testing strategy adequately addresses listed risks
- [ ] Newly-added code is easy to change
- [ ] Release note makes sense to a user of the library
- [ ] If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [ ] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

--------------------

[Python integration examples](https://github.com/DataDog/dd-trace-py/tree/main/ddtrace/contrib)
[Python integration docs](https://ddtrace.readthedocs.io/en/stable/contributing-integrations.html) 
[Running tests for python tracer, including integration tests](https://ddtrace.readthedocs.io/en/stable/contributing-testing.html)

## Testing Strategy 
Try functional testing(run an app with the build of the tracer with integration code) to see when things get executed and if the current wrapping point even makes sense(this hasn't been confirmed).

### CODE REVIEW

File: ddtrace/_trace/tracer.py [hunk 1]
```diff
@@ -549,9 +549,9 @@ def configure(
                 sync_mode=self._use_sync_mode(),
                 api_version=api_version,
                 # if apm opt out, neither agent or tracer should compute the stats
-                headers={"Datadog-Client-Computed-Stats": "yes"}
-                if (compute_stats_enabled or self._apm_opt_out)
-                else {},
+                headers=(
```
Comment (mabdinur) [inline ddtrace/_trace/tracer.py:552]: Do we need to make this change?

File: ddtrace/contrib/azure.functions/__init__.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+"""
```
Comment (mabdinur) [inline ddtrace/contrib/azure.functions/__init__.py:1]: Can we name this integration (and this package) `azure_functions`. This better aligns with out existing conventions

File: ddtrace/contrib/azure.functions/__init__.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+The foo integration instruments the bar and baz features of the
```
Comment (mabdinur) [inline ddtrace/contrib/azure.functions/__init__.py:2]: We need to update this docs page.

```suggestion
The azure function integration instruments the ..... and .... features of the
```

File: ddtrace/contrib/azure.functions/patch.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+from ddtrace import config
```
Comment (mabdinur) [inline ddtrace/contrib/azure.functions/patch.py:1]: Can you install hatch and then run the following command: `hatch run lint:fmt`.  This command will resolve the linting errors you're seeing in our tests

File: ddtrace/contrib/azure.functions/patch.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    "azure.functions",
+    {
+        "distributed_tracing": True,
+        "_default_service": schematize_service_name("azure.functions"),
+    },
+)
+
+
+def get_version():
+    return ""
```
Comment (mabdinur) [inline ddtrace/contrib/azure.functions/patch.py:17]: We need to implement this function. This should return the version of the azure function package

File: ddtrace/contrib/azure.functions/patch.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
+from ddtrace import config
+from ddtrace.internal.schema import schematize_service_name
+import azure.functions
+
+from ...internal.utils.wrappers import unwrap
+from ddtrace.vendor import wrapt
```
Comment (mabdinur) [inline ddtrace/contrib/azure.functions/patch.py:6]: Try using these wrappers instead.

```suggestion
from ddtrace.internal.wrapping import unwrap
from ddtrace.internal.wrapping import wrap
```

File: ddtrace/contrib/azure.functions/patch.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    # find function and wrap it
+    wrapt.wrap_function_wrapper("azure.functions.decorators.function_app", "Function.__init__", _wrapped_init)
+    pass
+
+
+def unpatch():
+    # Undo the monkey patching that patch() did here
+    if getattr(azure.functions, "_datadog_patch", False):
+        unwrap(azure.functions.decorators.function_app.Function, "__init__")
+        azure.functions._datadog_patch = False
```
Comment (mabdinur) [inline ddtrace/contrib/azure.functions/patch.py:35]: nit: this better aligns with how we handle unpatching in other integrations

```suggestion
    if getattr(azure.functions, "_datadog_patch", False):
        return

        unwrap(azure.functions.decorators.function_app.Function, "__init__")
        azure.functions._datadog_patch = False
```

File: releasenotes/notes/allow-mini-agent-to-run-on-all-hosting-plans-5e8df7330c108d71.yaml [hunk 8]
```diff
@@ -0,0 +1,4 @@
+
+features:
+  - |
+    tracing: This removes the restrictions on the tracer to only run the mini-agent on the consumption plan. The mini-agent now runs regardless of the hosting plan
```
Comment (mabdinur) [inline releasenotes/notes/allow-mini-agent-to-run-on-all-hosting-plans-5e8df7330c108d71.yaml:4]:
```suggestion
    azure_functions: Removes the restrictions on the tracer to only run the mini-agent on the consumption plan. The mini-agent now runs regardless of the hosting plan.
```

File: ddtrace/contrib/azure.functions/patch.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+
+def _wrapped_init(func, args, kwargs):
```
Comment (mabdinur) [inline ddtrace/contrib/azure.functions/patch.py:38]: What are you trying to patch here? Can you add a bit more logic here

**Review by mabdinur — COMMENTED** [review]
Can we split this up in to two PR. One PR for adding the azure function integration and the other for renaming `in_azure_function_consumption_plan` to `in_azure_function`

File: ddtrace/contrib/azure.functions/patch.py (hunk 7 above)
Comment (gord02) [reply ddtrace/contrib/azure.functions/patch.py:35]: Wouldn't returning first mean lines 36 and 37 wouldn't run?

**Pull request closed without merging**